### PR TITLE
Overridden functions should have the right prototype in an iframe without src (fix #1824)

### LINKED ISSUE #1824: Mootools errors from testcafe seem to cause a button to not function after clicking it
### Are you requesting a feature or reporting a bug?
Bug

### What is the current behavior?
I am unable to click a button, even if I pause the test and manually click the button. The test doesn't say the button can't be clicked but I can't progress in the application b/c clicking the button doesn't actually do anything. There are various mootools errors popping up in the console log that do not happen when you access the page without testcafe

This button gets clicked but doesn't actually function how it's supposed to. It would normally bring up a dialogue modal asking if you're sure you want to finish the practice test.

### What is the expected behavior?
The button should be able to be clickable and execute Its purpose and there shouldn't be any errors in the console log about mootools.

### How would you reproduce the current behavior (if this is a bug)?
Execute the test I have provided and try clicking the finish practice button at the end and notice that it will not do anything

#### Provide the test code and the tested page URL (if applicable)

Tested page URL:

**Note this url can't be directly accessed you would have to run the test or manually do what the test is doing to get here**

https://placeu.net/placementtest/take/1300/practice/5b85be7390b313f018ecf88e/1

Test code  **(I included a debug at the end so you can manually try to click and inspect the errors)**

```js
import { Selector }      from 'testcafe';

fixture `testcafe canvas`
    .page `https://webassign.instructure.com/login/canvas`;

const medwait            = 5000
const longwait           = 15000;
const course             = Selector('.ic-DashboardCard__header-title');
const ltiLink            = Selector('.item_link');
const loadBtn            = Selector('.btn[type="submit"]');
const startPracticeBtn   = Selector('button[wa-qa="placement-practice-button"]')
const confirmPracticeBtn = Selector('start-practice-confirmation button.wa-button-primary');
const finishPracticeBtn  = Selector('wa-button[on-click="$ctrl.finishPractice()"] button.wa-submit-test-btn');
const spinner            = Selector('.md-spinner-wrapper');

test('Instructor', async t => {
    await t
        .typeText(Selector('#pseudonym_session_unique_id'), 'testcafestudent01@testing123.com')
        .typeText(Selector('#pseudonym_session_password'), 'Password1')
        .click(Selector('#login_form .Button--login'))
        .expect(course.exists).ok({timeout: medwait})
        .click(course)
        .expect(ltiLink.exists).ok({timeout: medwait})
        .click(ltiLink)
        .expect(loadBtn.exists).ok({timeout: medwait})
        .click(loadBtn)
        .expect(startPracticeBtn.exists).ok({timeout: medwait})
        .click(startPracticeBtn)
        .expect(confirmPracticeBtn.exists).ok({timeout: medwait})
        .click(confirmPracticeBtn)
        .expect(spinner.exists).notOk({ timeout: longwait })
        .expect(finishPracticeBtn.exists).ok({timeout: medwait})
        .click(finishPracticeBtn).debug();
});
```

### Specify your

* operating system: Mac Osx Sierra
* testcafe version: 0.21.1
* node.js version: 8.2.1

### PR DESCRIPTION
**Pull request by LavrovArtem**

### CODE REVIEW

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

**Comment by Farfurix** [comment]
@LavrovArtem https://github.com/DevExpress/testcafe-hammerhead/pull/1837 (Fix failed Internet Explorer 11 tests (NetworkError occurs after some Windows 10 update, iframe without src case)) is merged, need to rebase.

File: src/client/sandbox/index.js [hunk 1]
```diff
@@ -151,7 +151,9 @@ export default class Sandbox extends SandboxBase {
         urlResolver.init(this.document);
 
         // NOTE: Eval Hammerhead code script.
-        this.iframe.on(this.iframe.EVAL_HAMMERHEAD_SCRIPT_EVENT, e => initHammerheadClient(e.iframe.contentWindow, true));
+        this.iframe.on(this.iframe.EVAL_HAMMERHEAD_SCRIPT_EVENT, e => {
+            e.iframe.contentWindow.eval(`(${ initHammerheadClient.toString() })();//# sourceURL=hammerhead.js`);
```
Comment (miherlosev) [inline src/client/sandbox/index.js]: Is instruction `//# sourceURL=hammerhead.js` necessary?

File: src/client/utils/html.js [hunk 2]
```diff
@@ -136,13 +131,58 @@ export function cleanUpHtml (html) {
         let changed = false;
 
         find(container, STORED_ATTRS_SELECTOR, el => {
-            for (const { attr, storedAttr } of ATTRS_DATA_FOR_CLEANING) {
-                if (el.hasAttribute(attr) && el.hasAttribute(storedAttr))
-                    nativeMethods.setAttribute.call(el, attr, nativeMethods.getAttribute.call(el, storedAttr));
-                else if (attr === 'autocomplete')
-                    nativeMethods.removeAttribute.call(el, attr);
+            const urlAttr = domProcessor.getUrlAttr(el);
```
Comment (miherlosev) [inline src/client/utils/html.js]: Let's try to refactor it in this manner:
```js
find(container, STORED_ATTRS_SELECTOR, el => { 
  ....
  cleanUpStoredAttr(...);
  cleanUpAutocompleteMarker();
  cleanUpIframeSandboxAttr();
...etc.
```

File: src/client/utils/url-resolver.js [hunk 3]
```diff
@@ -53,8 +54,10 @@ export default {
             // NOTE: It looks like a Chrome bug: in a nested iframe without src (when an iframe is placed into another
             // iframe) you cannot set a relative link href while the iframe loading is not completed. So, we'll do it with
             // the parent's urlResolver Safari demonstrates similar behavior, but urlResolver.href has a relative URL value.
-            const needUseParentResolver = url && isIframeWithoutSrc && window.parent && window.parent.document &&
-                                          (!href || href.indexOf('/') === 0);
+            const isIframeWithUnresolvedUrl = url && window.parent && window.parent !== window &&
```
Comment (miherlosev) [inline src/client/utils/url-resolver.js]: It looks like an intersection or bad variable naming.
You 3 times check that current window is iframe. 
Try to refactor.

File: src/client/utils/url.js [hunk 4]
```diff
@@ -23,9 +23,17 @@ export function getProxyUrl (url, opts) {
         return url;
 
     /*eslint-disable no-restricted-properties*/
-    const proxyHostname       = opts && opts.proxyHostname || location.hostname;
-    const proxyPort           = opts && opts.proxyPort || location.port.toString();
-    const proxyServerProtocol = opts && opts.proxyProtocol || location.protocol;
+    let locationWindow = window;
+    let proxyLocation  = locationWindow.location;
+
+    while (!proxyLocation.hostname) {
```
Comment (miherlosev) [inline src/client/utils/url.js]: Move to a separate function and add comment why it is necessary.

File: src/client/sandbox/index.js (hunk 1 above)
Comment (LavrovArtem) [reply src/client/sandbox/index.js]: It needs for debugging purposes.
image

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

File: src/client/sandbox/index.js (hunk 1 above)
Comment (LavrovArtem) [reply src/client/sandbox/index.js]: image

**Pull request merged**